src/translator/service.h

```diff
@@ -9,31 +9,71 @@
 #include "quality_estimator.h"
 #include "response.h"
 #include "response_builder.h"
+#include "tensors/tensor_allocator.h"
 #include "text_processor.h"
 #include "threadsafe_batching_pool.h"
 #include "translation_model.h"
 #include "translator/parser.h"
 #include "vocabs.h"
-
 namespace marian {
 namespace bergamot {
 
+class Workspace {
+ public:
+  Workspace(size_t deviceId, size_t workspaceSizeInMB) : deviceId_(deviceId) {
+    // We'll eventually get rid of this, but proof of concept faster?
+    auto graph = New<ExpressionGraph>(/*inference=*/true);  // set the graph to be inference only
+    const std::string precision{"float32"};
+    graph->setDefaultElementType(typeFromString(precision));
+    marian::DeviceId device(deviceId, DeviceType::cpu);
+    graph->setDevice(device);
+    graph->getBackend()->configureDevice(horribleOptionsHack());
+    Ptr<Backend> backend = graph->getBackend();
+
+    tensors_ = New<TensorAllocator>(backend);
+    tensors_->reserve(workspaceSizeInMB);
+
+    cache_ = New<TensorAllocator>(backend);
+  }
+
+  Ptr<TensorAllocator> tensors() { return tensors_; }
+  Ptr<TensorAllocator> cache() { return cache_; }
+
+  size_t id() { return deviceId_; }
+
+ private:
+  Ptr<TensorAllocator> tensors_{nullptr};
+  Ptr<TensorAllocator> cache_{nullptr};
```

Comment:
> 1. How does this relate to what is inside the workspace?
> 2. How does this relate to what is inside longterm/cache?


A workspace strictly corresponds to a `TensorAllocator`. There are two in `ExpressionGraph`, one for `tensors_`, one for `cache_`.

```
ExpressionGraph contains Tensors.
Tensors := (
	tensors_ : TensorAllocator, 
	cache_ : TensorAllocator, 
	longterm_: unordered_map<node->hash(), Tensor>, 
	shorterm_: unordered_map<node->hash(), Tensor>
)
```

It is `tensors_` and `cache_` we lift (from being hidden behind multiple layers in a graph) and place with Service/worker. The graph can now accept these as a parameter to share a common tensor-allocation backend.

`longterm_`, `shortterm_` are effectively namespaces/environments for models and hence isolated (no sharing). This prevents any conflicts in names/hashing same tensor contents belonging to different models into same namespace/env.